### PR TITLE
Auto-indexing for `SELECT`s in `sqllogictest`

### PR DESCRIPTION
**Pull request by vmarcos**
This PR introduces automatic creation of indexed views and redundant execution for `SELECT` queries in `sqllogictest` by the new argument `--auto-index-selects`. The outcome for the vanilla execution of the query is contrasted with the outcome of creating an indexed view with the same `SELECT` query and selecting from this view. This feature will enable us to ensure via SLT-based testing that optimizations that change the planning and execution for one-shot `SELECT`s can be validated, in terms of produced results, against the standard optimization and execution pipeline for incremental view maintenance.
 
The feature is intended at allowing us to reuse existing corpora of SLTs while validating both view-based and one-shot query processing. It is an explicit non-goal to achieve 100% reuse, i.e., there will be SLTs where the feature will generate diverging results for view-based and one-shot query processing due to difficulties in extracting semantic information from the SQL syntax. For example, `SELECT`s allow for ambiguity in output column names, while views do not. Additionally, `ORDER BY` clauses in views are ignored, but are tricky to reliably reintroduce in the outer query in view-based execution. Only best-effort remediation of such potential issues is implemented by this PR. We aim at maximizing reuse of SLT corpora at manageable implementation effort for the feature. 

Advances #18734.

### Motivation

  * This PR adds a known-desirable feature. https://github.com/MaterializeInc/materialize/issues/18734

### Tips for reviewer

I would suggest to consider the commits individually. Some small code organization was necessary prior to implementing the bulk of the feature. The code organization is performed in the first two commits, while the core of the feature is in the third commit. The fourth and sixth commits provide derivations of number of attributes and `ORDER BY` clauses. The fifth commit introduces unit tests and adds one SLT to the fast SQL logic tests being run with this feature. The seventh commit is code reorg / movement, so the diff ended up unfortunately a bit hard to parse (sorry!). The eighth commit refines the rewrite strategy by attempting to not use analysis information about the number of attributes at first (thus we can reuse `ORDER BY` with expressions on attributes projected, for example), but falling back in using the information and attempting a more involved `ORDER BY` / `SELECT` projection rewrite in case of error. Note that we needed to exclude `mz_sqllogictest` from Miri, as the new unit test raises the unavailability of `rust_psm_stack_pointer()` due to heavier use of the parser.

I tried to run several SLTs with the new option `--auto-index-selects`. Here are a few where we got clean runs locally:

- `test/sqllogictest/aggregates.slt` 
    - This test is included in this PR in the fast SQL logic tests along with `--auto-index-selects`. In addition, a unit test is included. Other tests below could also be considered to further increase coverage of this PR. I am open to suggestions that strike a good trade-off with increased CI time. 
- `test/sqllogictest/arrays.slt`
- `test/sqllogictest/array_subquery.slt`
- `test/sqllogictest/float.slt`
- `test/sqllogictest/list_subquery.slt`
- `test/sqllogictest/numeric.slt`
- `test/sqllogictest/order_by.slt`
- `test/sqllogictest/scalar_subqueries_select_list.slt`
- `test/sqllogictest/session-window-wmr.slt`
- `test/sqllogictest/topk.slt`
- `test/sqllogictest/unsigned_int.slt`
- `test/sqllogictest/with_mutually_recursive.slt`
- `test/sqllogictest/cockroach/distinct_on.slt`
- `test/sqllogictest/cockroach/with.slt`
- `test/sqllogictest/cockroach/limit.slt`
- `test/sqllogictest/cockroach/subquery_correlated.slt`
- `test/sqllogictest/cockroach/order_by.slt`
- `test/sqllogictest/postgres/subselect.slt`

The SLTs above were selected based on including min/max or top-k patterns plus being focused on query execution, not planning. The list is not exhaustive. There were some SLTs where the rewrite strategies above failed. A few prominent reasons are:

1. The test contains a `SELECT` case that forces expected columns with ambiguous names (e.g., `test/sqllogictest/cockroach/aggregate.slt`, `test/sqllogictest/postgres/union.slt`).
2. The test contains a `SELECT` case that uses wildcards and subqueries, defeating the currently implemented analysis and rewrites (e.g., `test/sqllogictest/window_funcs.slt`, `test/sqllogictest/cockroach/union.slt`).
3.  The test contains a `SELECT` with an unmaterializable function (e.g., `test/sqllogictest/list.slt`, `test/sqllogictest/dates-times.slt`, `test/sqllogictest/pg_catalog_class.slt`) or on a relation where the dependency is deemed unstable (e.g. `test/sqllogictest/audit_log.slt`).

Some of the cases above could be addressed by, e.g., breaking down the tests into the problematic queries (usually few) and the rest of the test. I propose that we iterate over potential strategies to address the limitations above in further PRs, if necessary, to avoid extending the scope of this PR even further.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/vmarcos/materialize/blob/main/doc/developer/design/20230421_stabilize_monotonic_select.md).
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label. N/A
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).N/A
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note): N/A

### CODE REVIEW

**Review by def- — APPROVED** [review]
This is really cool. I'll look into the long SLT files and check which ones we can enable `--auto-index-selects` for: https://buildkite.com/materialize/sql-logic-tests/builds/5572 My initial gut feeling would be to try to enable it in all SLT files that work for the daily long runs. For the short SLT, we are currently at ~10 minutes and it runs in each PR, so maybe only a few key files.

File: ci/test/slt-fast.sh [hunk 1]
```diff
@@ -159,4 +159,10 @@ tests=(
     test/sqllogictest/postgres/pgcrypto/*.slt
 )
 
+tests_with_views=(
+    test/sqllogictest/aggregates.slt
```
Comment (def-) [inline ci/test/slt-fast.sh:163]: Does it make sense to still run aggregates.slt without `--auto-index-selects` in addition?

**Comment by vmarcos** [comment]
> This is really cool. I'll look into the long SLT files and check which ones we can enable `--auto-index-selects` for: https://buildkite.com/materialize/sql-logic-tests/builds/5572 My initial gut feeling would be to try to enable it in all SLT files that work for the daily long runs. For the short SLT, we are currently at ~10 minutes and it runs in each PR, so maybe only a few key files.

Awesome! I looked into the long SLT build output linked above, which contained a number of failures. Unfortunately, the `--auto-index-selects` option does not play well with the `--auto-transactions` option. This is because we need to execute DDL to create the view / index for each `SELECT`, and if we are in a transaction an abort ensues. Is it realistic to run the long SLT files without `--auto-transactions`? Alternatively, we might need to choose a subset for these tests as well.

**Comment by def-** [comment]
I have seen that as well. Without auto-transactions the runtime would go from ~6 hours to 20 hours, so probably not.

**Review by def- — APPROVED** [review]
The tests are still fast enough at 10 min, so good to go from my side.

**Review by def- — COMMENTED** [review]
One thing I noticed in https://buildkite.com/materialize/sql-logic-tests/builds/5573#0188d842-25a7-46fe-b6aa-691611736b61 is that the sqllogictest runner can crash if the materialized view can't be created: `bin/sqllogictest -- --auto-index-selects -v -v --no-fail test/sqllogictest/cockroach/tuple.slt`:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "In test/sqllogictest/cockroach/tuple.slt:\ndb error: ERROR: cannot materialize call to current_timestamp\nDETAIL: See: https://materialize.com/docs/sql/functions/now_and_mz_now/\nHINT: Try using `mz_now()` here instead."', /home/deen/git/materialize2/src/sqllogictest/src/runner.rs:1644:14
```
I'm not sure if this should be handled better. I'll ignore the affected files for now.

File: ci/test/slt-fast.sh (hunk 1 above)
Comment (def-) [reply ci/test/slt-fast.sh:163]: Another thing is, maybe we should turn around the logic here and have `tests_without_views`, if it's not too many, so that new tests are run with --auto-index-selects by default.

File: src/sqllogictest/src/runner.rs [hunk 2]
```diff
@@ -1739,6 +1901,298 @@ impl<'a> RewriteBuffer<'a> {
...
+    let create_view = AstStatement::<Raw>::CreateView(CreateViewStatement {
+        if_exists: IfExistsBehavior::Error,
+        temporary: false,
+        definition: ViewDefinition {
+            name: name.clone(),
+            columns: columns.clone(),
+            query,
+        },
+    })
+    .to_ast_string_stable();
```
Comment (def-) [inline src/sqllogictest/src/runner.rs:1996]: I'm wondering if it would be helpful to print out the views/indexes created here too when `-v -v` is used.

**Comment by def-** [comment]
With this example:
```
mode cockroach

query T
SELECT INTERVAL '1:2:31.23456789' MINUTE TO SECOND(2);
----
00:02:31.23
```
`sqllogictest --auto-index-selects` generates this:
```
CREATE VIEW "v4a9fabbde3f14b33afcb0d8b50d57e49" AS SELECT INTERVAL '1:2:31.23456789' SECOND (2);
CREATE DEFAULT INDEX ON "v4a9fabbde3f14b33afcb0d8b50d57e49";
SELECT * FROM "v4a9fabbde3f14b33afcb0d8b50d57e49";
```
So it loses the "MINUTE TO" part, thus leading to inconsistent result.

**Comment by vmarcos** [comment]
> One thing I noticed in https://buildkite.com/materialize/sql-logic-tests/builds/5573#0188d842-25a7-46fe-b6aa-691611736b61 is that the sqllogictest runner can crash if the materialized view can't be created: `bin/sqllogictest -- --auto-index-selects -v -v --no-fail test/sqllogictest/cockroach/tuple.slt`:
> 
> ```
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "In test/sqllogictest/cockroach/tuple.slt:\ndb error: ERROR: cannot materialize call to current_timestamp\nDETAIL: See: https://materialize.com/docs/sql/functions/now_and_mz_now/\nHINT: Try using `mz_now()` here instead."', /home/deen/git/materialize2/src/sqllogictest/src/runner.rs:1644:14
> ```
> 
> I'm not sure if this should be handled better. I'll ignore the affected files for now.

I just added a commit now that should improve the error handling when indexing the automatically created view. So now we should see this error become a test outcome.

File: src/sqllogictest/src/runner.rs (hunk 2 above)
Comment (vmarcos) [reply src/sqllogictest/src/runner.rs:2035]: Good idea, I just added some support now to output the automatically generated view creation, indexing, querying, and dropping statements when the verbosity level is set high.

**Comment by vmarcos** [comment]
> With this example:
> 
> ```
> mode cockroach
> 
> query T
> SELECT INTERVAL '1:2:31.23456789' MINUTE TO SECOND(2);
> ----
> 00:02:31.23
> ```
> 
> `sqllogictest --auto-index-selects` generates this:
> 
> ```
> CREATE VIEW "v4a9fabbde3f14b33afcb0d8b50d57e49" AS SELECT INTERVAL '1:2:31.23456789' SECOND (2);
> CREATE DEFAULT INDEX ON "v4a9fabbde3f14b33afcb0d8b50d57e49";
> SELECT * FROM "v4a9fabbde3f14b33afcb0d8b50d57e49";
> ```
> 
> So it loses the "MINUTE TO" part, thus leading to inconsistent result.
> 

AFAICT, this looks like a bug in the code that formats the `IntervalValue`: https://github.com/MaterializeInc/materialize/blob/ad6d3197ef88f4d067c40532bc2eb810008ace5b/src/sql-parser/src/ast/defs/value.rs#L78-L126
Note above that the `Interval` variant is matched twice. If some `fsec_max_precision` is given, the first arm is taken, ignoring the precision fields. Commenting out the first match arm fixes the problem locally for me. However, I think this should probably be raised to someone in Surfaces to evaluate, since I don't know if the formatted values are saved anywhere (e.g. in the stash).

**Comment by def-** [comment]
Great, I have opened an issue for surfaces team: https://github.com/MaterializeInc/materialize/issues/20110 So shouldn't block this.

**Review by def- — APPROVED** [review]
I have added my change here to run this in full SLT runs too, also triggered a final run for validation: https://buildkite.com/materialize/sql-logic-tests/builds/5583
But should be good to submit.

File: ci/test/slt-fast.sh [hunk 3]
```diff
@@ -19,6 +19,203 @@ tests=(
...
+    test/sqllogictest/transform/reduce_fusion.slt
+    test/sqllogictest/transform/reduction_pushdown.slt
+    test/sqllogictest/transform/redundant_join.slt
+    test/sqllogictest/transform/relation_cse.slt
+    test/sqllogictest/transform/scalability.slt
+    test/sqllogictest/transform/scalar_cse.slt
+    test/sqllogictest/transform/threshold_elision.slt
+    test/sqllogictest/transform/topk.slt
+    test/sqllogictest/transform/union.slt
+    test/sqllogictest/transform/union_cancel.slt
```
Comment (def-) [inline ci/test/slt-fast.sh:218]: That was more than expected

File: ci/test/slt-fast.sh (hunk 1 above)
Comment (vmarcos) [reply ci/test/slt-fast.sh]: This is a great suggestion. I have added a commit now that runs the fast SQL logic tests with `--auto-index-selects` by default, following a similar structure to your commit for the slow SQL logic tests.

**Review by mjibson — APPROVED** [review]

**Comment by vmarcos** [comment]
> Based on https://buildkite.com/materialize/sql-logic-tests/builds/5578#0188de5e-7ce2-4a67-ae2c-00d62ca61190 I now have a bunch of failures which I'm ignoring: `grep "actually from indexed" run.log|grep -v "specified more than once"|grep -v "does not exist"|grep -v "COLLATE not yet supported"|grep -v "cannot create view with unstable dependencies" | grep -v "appears more than once" | grep -v "cannot use wildcard expansions or NATURAL JOINs" | grep -v "current transaction is aborted, commands ignored until end of transaction block" | grep -v "Evaluation error:" | grep -v "UnexpectedPlanSuccess"`. Could we maybe detect some of them in SLT and then not emit views for them or would that be too complex?
> 

It is not impossible, but a good fix for the issue probably exceeds the scope of this PR. I am filing a follow-up issue so that we do not forget: https://github.com/MaterializeInc/materialize/issues/20146.

My reasoning is as follows. The PR abides by two important principles for `--auto-index-selects`:
1. If we are able to parse a `SELECT`, then we attempt an indexed view execution for it.
2. All errors and inconsistent outcomes are raised, so that we get full visibility on all the semantic differences (or bugs!) that might occur between executing a one-shot `SELECT` or executing the same statement through an indexed view.

If we ignore some errors and/or parse some queries and decide not to run them as indexed views, we'd be violating one of the principles above. I think it is reasonable to do so in a controlled fashion, but it would be good to think of a mechanism by which we'd not lose observability of the issues. For example, we could introduce a warning outcome in `sqllogicttest` that would be emitted under the `--auto-index-selects` option. The warning outcome would raise the issue, but not fail the SLT. It might even be useful to have such a warning outcome outside of the `--auto-index-selects` context too.

**Review by def- — APPROVED** [review]
Good to submit from my side, great work @vmarcos !

**Comment by vmarcos** [comment]
Thanks for the reviews and contributions!!
